ImageData using storageFormat

squelart opened the pull request:
#### 6f2ac99d8410ead797a23036ba1b442669f61fa7
<pre>
ImageData using storageFormat
<a href="https://bugs.webkit.org/show_bug.cgi?id=284004">https://bugs.webkit.org/show_bug.cgi?id=284004</a>
<a href="https://rdar.apple.com/problem/140877760">rdar://problem/140877760</a>

Reviewed by Said Abou-Hallawa.

ImageData now supports:
- The IDL property `storageFormat` of type `ImageDataStorageFormat`, which
  reflects the ImageDataSettings.storageFormat value given during construction.
- The IDL property `data` can now be `Uint8ClampedArray` or `Float16Array`,
  depending on the storageFormat.

The main difference for now, is that
`context.createImageData(w, h, { storageFormat: &quot;float16&quot; })`
should correctly output an ImageData with data of type Float16Array.

Otherwise most uses of ImageData, like in CanvasRenderingContext2DBase, enforce
the use of Uint8ClampedArray, so there shouldn&apos;t be any visible changes in most
cases, including when requesting
`context.getImageData(w, h, { storageFormat: &quot;float16&quot; })`
(to be implemented in future PRs).

* LayoutTests/fast/canvas/imagedata-storageformat-enabled-expected.txt:
* LayoutTests/fast/canvas/imagedata-storageformat-enabled.html:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/bindings/js/JSImageDataCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::CloneDeserializer::readTerminal):
* Source/WebCore/html/ImageData.cpp:
(WebCore::computeDataSize):
(WebCore::computeStorageFormat):
(WebCore::ImageData::create):
(WebCore::ImageData::pixelBuffer const):
(WebCore::ImageData::createUninitialized): Deleted.
(WebCore::ImageData::clone const): Deleted.
* Source/WebCore/html/ImageData.h:
(WebCore::ImageData::create):
(WebCore::ImageData::data const):
(WebCore::ImageData::storageFormat const):
* Source/WebCore/html/ImageData.idl:
* Source/WebCore/html/ImageDataArray.cpp:
(isType):
(WebCore::visitArrayBufferView):
(WebCore::ImageDataArray::ImageDataArray):
(WebCore::fillTypedArray):
(WebCore::ImageDataArray::tryCreate):
(WebCore::ImageDataArray::isSupported):
(WebCore::ImageDataArray::storageFormat const):
(WebCore::convertToTypedArray):
(WebCore::ImageDataArray::asUint8ClampedArray const):
(WebCore::ImageDataArray::asFloat16Array const):
(WebCore::ImageDataArray::length const):
(WebCore::ImageDataArray::copyToJSONArray const):
(WebCore::ImageDataArray::operator Variant const):
* Source/WebCore/html/ImageDataArray.h:
(WebCore::ImageDataArray::isSupported):
(WebCore::ImageDataArray::tryCreate):
(WebCore::ImageDataArray::arrayBufferView const):
(WebCore::ImageDataArray::protectedArrayBufferView const):
(WebCore::ImageDataArray::byteLength const):
(WebCore::ImageDataArray::isDetached const):
(WebCore::ImageDataArray::span const):
(WebCore::ImageDataArray::ImageDataArray): Deleted.
(WebCore::ImageDataArray::length const): Deleted.
(WebCore::ImageDataArray::get const): Deleted.
(WebCore::ImageDataArray::data const): Deleted.
* Source/WebCore/html/ImageDataStorageFormat.h:
(WebCore::toPixelFormat):
(WebCore::toImageDataStorageFormat):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::createImageData const):
(WebCore::CanvasRenderingContext2DBase::cacheImageDataIfPossible):
(WebCore::CanvasRenderingContext2DBase::makeImageDataIfContentsCached const):
(WebCore::CanvasRenderingContext2DBase::getImageData const):
(WebCore::initializeEmptyImageData): Deleted.
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::texImageSource):
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::buildArrayForImageData):

Canonical link: <a href="https://commits.webkit.org/287598@main">https://commits.webkit.org/287598@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/227d673009541c0dd1ba4930e17ee22cb2810df7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80215 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59217 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33628 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84732 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31191 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82326 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68278 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7516 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62704 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20517 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83284 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52774 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73051 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43012 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50103 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27206 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29652 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71232 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27716 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86165 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7435 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5263 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70977 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7610 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68891 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70219 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17484 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14213 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13162 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7399 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12916 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7238 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10758 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9043 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -4966,10 +4966,12 @@ class CloneDeserializer : public CloneBase {
+                    memcpy(array.data(), bufferStart, length);
+                else
+                    array.zeroFill();
+            });
```
This can be handled by `ImageDataArray`. I think it is better to hide the fact that `ImageDataArray` is generic `ImageData` array and can hold different types of array elements.
```
result.returnValue()->data().copyData(bufferStart, length);
```

shallawa commented on Source/WebCore/html/ImageData.cpp:
```diff
@@ -153,13 +177,16 @@ ImageData::~ImageData() = default;
-    return ImageData::create(m_size, Uint8ClampedArray::create(m_data.data(), m_data.length()), m_colorSpace);
+    return m_data.visit([self = Ref(*this)](auto& array) {
+        return ImageData::create(self->m_size, array.create(array.data(), array.length()), self->m_colorSpace);
+    });
```
This can be 
```
return ImageData::create(m_size, m_data.clone(), m_colorSpace);
```

squelart replied:
This function was never used! So I've just removed it. We can re-implement clone if&whenever needed...

shallawa commented on Source/WebCore/html/ImageDataArray.h:
```diff
@@ -28,34 +28,147 @@
-    auto data() const
+    // Needed by `toJS<IDLUnion<IDLUint8ClampedArray, ...>, const ImageDataArray&>()`
+    using Variant = std::variant<RefPtr<JSC::Uint8ClampedArray>, RefPtr<JSC::Float16Array>>;
+    operator Variant() const
```
What is this casting operator is supposed to do? I feel it is a little bit weird to say
```
auto array = static_cast<ImageDataArray:: Variant>imageData.data();
```

squelart replied:
As the comment above says, it's needed by `toJs`. Without it, we get this compilation error:
> no viable conversion from 'const WebCore::ImageDataArray' to 'const ImplementationType' (aka 'const std::variant<RefPtr<JSC::GenericTypedArrayView<JSC::Uint8ClampedAdaptor>>, RefPtr<JSC::GenericTypedArrayView<JSC::Float16Adaptor>>>')

I think it's because in ImageData.idl we have `typedef (Uint8ClampedArray or Float16Array) ImageDataArray;`, this eventually generates JS binding code that expects ImageData.data to be an `IDLUnion` which corresponds to a C++ `std::variant`.

There may be other ways to deal with this, through custom JS converters, but here it seemed easy enough to just provide the `std::variant` that generated JS bindings want.

- Can we keep this for now, and revisit later if you think it would be worth it?
- Should I add more details in the comment above?
- I could potentially hide it in the `private:` section, with the appropriate `friend`ship to let `toJS` use it; this would avoid random uses of this one strange function. What do you think?

shallawa commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -2510,10 +2512,7 @@ RefPtr<ByteArrayPixelBuffer> CanvasRenderingContext2DBase::cacheImageDataIfPossi
-        if (imageData)
-            imageData->data().zeroFill();
-        return imageData;
+        return ImageData::createZeroFilled(sourceRect.size(), colorSpace);
```
Maybe you can leave this as is if you implement
```
ImageDataArray::zeroFill()
```

squelart replied:
After further reflection, I've decided to keep my changes, because I think it's a safer API to have:
one function `createZeroFilled()` that returns a fully-initialized object;
rather than a combination of `create()` returning an unfinished ImageData object that should not be used and then a `data()->zeroFill` that magically does the right thing to finalize the object.
Also, it makes for a nice one-liner `return`.

But please let me know if you disagree, and you think I really should keep the Create+zeroFill combination (and could you please explain why you'd prefer that?).

shallawa commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -2431,7 +2431,9 @@ bool CanvasRenderingContext2DBase::needsPreparationForDisplay() const
-    imageData.data().zeroFill();
+    imageData.data().visit([](auto& array) {
+        array.zeroFill();
+    });
```
Maybe you can leave this as is if you implement
```
ImageDataArray::zeroFill()
```

squelart replied:
I've just removed this function, see comment below...

shallawa commented on Source/WebCore/inspector/InspectorCanvas.cpp:
```diff
@@ -1393,14 +1393,29 @@ Ref<JSON::ArrayOf<JSON::Value>> InspectorCanvas::buildArrayForCanvasPattern(cons
+    for (size_t i = 0; i < imageDataArray.length(); ++i)
+        jsArray->addItem(imageDataArray.item(i));
+    return jsArray;
+}
```
I think it is better to hide these details about `ImageDataArray` from the caller. The caller should not have to implement these casting operator helpers or the template functions.

shallawa commented on Source/WebCore/inspector/InspectorCanvas.cpp:
```diff
@@ -1393,14 +1393,29 @@ Ref<JSON::ArrayOf<JSON::Value>> InspectorCanvas::buildArrayForCanvasPattern(cons
+        array->addItem(copyImageDataToJSONArray(imageData));
+    });
     array->addItem(imageData.width());
     array->addItem(imageData.height());
```
Can't this be just:
```
auto array = imageData.data().copyToJSONArray();
array->addItem(imageData.width());
array->addItem(imageData.height());
```

shallawa commented on Source/WebCore/html/ImageDataArray.h:
```diff
@@ -28,34 +28,147 @@
+    auto visit(F&& f) const
+    {
+        return visit(m_bufferView.get(), std::forward<F>(f));
+    }
```
I would love to see all the visit methods be private to this class. The caller should not have to deal with the casting or enumeration. The caller side should not be affected much by this change.

shallawa commented on Source/WebCore/bindings/js/JSImageDataCustom.cpp:
```diff
@@ -43,10 +43,12 @@ JSValue toJSNewlyCreated(JSGlobalObject* lexicalGlobalObject, JSDOMGlobalObject*
-    wrapper->putDirect(vm, dataName, toJS(lexicalGlobalObject, globalObject, data), PropertyAttribute::DontDelete | PropertyAttribute::ReadOnly);
+    wrapper->putDirect(vm, dataName, data.visit([&](auto& array) {
+        return toJS(lexicalGlobalObject, globalObject, array);
+    }), PropertyAttribute::DontDelete | PropertyAttribute::ReadOnly);
```
I do not think this change is needed. `toJS()` above calls this function:
```
inline JSC::JSValue toJS(JSC::JSGlobalObject* lexicalGlobalObject, JSC::JSGlobalObject* globalObject, JSC::ArrayBufferView& view)
{
    return view.wrap(lexicalGlobalObject, globalObject);
}
```
So whether you pass `Uint8ClampedArray` or `Float16Array`, it will be casted to the base class `ArrayBufferView`. So I think this can be changed to 
```
auto& data = imageData->data().getArrayBufferView();
wrapper->putDirect(vm, dataName, toJS(lexicalGlobalObject, globalObject, data), PropertyAttribute::DontDelete | PropertyAttribute::ReadOnly);
```

shallawa commented on Source/WebCore/html/ImageDataArray.h:
```diff
@@ -28,34 +28,147 @@
+    Ref<JSC::Uint8ClampedArray> asUint8ClampedArray() const;
+    Ref<JSC::Float16Array> asFloat16Array() const;
+
+    JSC::ArrayBufferView& getArrayBufferView() const { return m_bufferView.get(); }
```
Can't `getArrayBufferView()` be just `data()`?

squelart replied:
I guess I was tired of all the `data` functions, giving us expressions like `blah->data()->data()->data()` 😄
Anyway, I've just removed this one, it was only used once in this same file, so no real need.

(And thanks for the thorough review, I'm working on addressing your other more important comments now...)

squelart replied:
Trying to remove the `visit` in JSImageDataCustom.cpp (as you suggested), I now need this getter-returning-ArrayBufferView back!
So the caller would be `toJS(..., arrayBufferView())` (I've removed the 'get').

Are you sure you want "data"? It would make the caller look like: `toJS(..., data.data())`.
Anyway, I'll try to rework some names for clarity in that JSImageDataCustom.cpp, please have a look when I post the update...

shallawa commented on Source/WebCore/html/ImageDataArray.h:
```diff
@@ -28,34 +28,147 @@
+            }
+        }
+        return array;
+    }
```
Why do we keep these long functions in the header file? This makes the interface of the class harder to see.

shallawa commented on Source/WebCore/bindings/js/JSImageDataCustom.cpp:
```diff
@@ -40,13 +40,13 @@ using namespace JSC;
     // FIXME: Adopt reportExtraMemoryVisited, and switch to reportExtraMemoryAllocated.
     // https://bugs.webkit.org/show_bug.cgi?id=142595
-    vm.heap.deprecatedReportExtraMemory(data.length());
+    vm.heap.deprecatedReportExtraMemory(imageDataArray->byteLength());
```
This is hard to read. It requires me to look at the implementation of `ImageDataArray` to understand why the arrow operator is used here although `imageDataArray` is a reference. Then I realized that it is equivalent to
```
imageDataArray.arrayBufferView().byteLength()
```
Please add `ImageDataArray::byteLength()` and use the dot operator here.

shallawa commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -1871,13 +1871,13 @@ class CloneSerializer : public CloneBase {
                 }
                 write(dataLength);
-                write(data->data().span());
+                write(data->data()->span());
```
Please add `ImageDataArray::span()`

shallawa commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -4960,16 +4960,12 @@ class CloneDeserializer : public CloneBase {
                 return jsNull();
 
-            auto result = ImageData::createUninitialized(width, height, resultColorSpace);
+            auto result = ImageData::createFromBufferOrZero(width, height, resultColorSpace, bufferStart, length);
```
I think this should be just `ImageData::create()`. The extra parameters `bufferStart, length` are sufficient to tell what this function is going to do. `ImageData` has similar `create` methods which initializes the object with certain data.

shallawa commented on Source/WebCore/html/ImageBitmap.cpp:
```diff
@@ -896,7 +896,7 @@ void ImageBitmap::createCompletionHandler(ScriptExecutionContext& scriptExecutio
     // 6.2. If IsDetachedBuffer(buffer) is true, then return p rejected with an
     //      "InvalidStateError" DOMException.
-    if (imageData->data().isDetached()) {
+    if (imageData->data()->isDetached()) {
```
Please add `ImageDataArray::isDetached()`.

shallawa commented on Source/WebCore/html/ImageData.cpp:
```diff
@@ -36,9 +36,24 @@
+        return PixelFormat::RGBA16F;
+#else
+        break;
+#endif
```
This is a little bit strange. Why do enable `ImageDataStorageFormat::Float16` unconditionally but we enable `PixelFormat::RGBA16F` only if `#if HAVE(HDR_SUPPORT)` is true? Maybe we need to enable  `PixelFormat::RGBA16F` unconditionally later in a future patch.

squelart replied:
I'd love `PixelFormat::RGBA16F` to be unconditional! 😄 But maybe that would be hard to implement on some non-Cocoa platforms?
I did make `ImageDataStorageFormat::Float16` unconditional because I think:
- The proposal suggested that it should be available everywhere.
- It's possible to implement it without backed support, with `put/getImageData` doing conversions on the fly.
- IIRC I initially made it conditional at the IDL level, and got compilation errors in generated code, and didn't pursue it further.

But we could decide to keep both in sync if that's preferable. To be discussed...

shallawa commented on Source/WebCore/html/ImageData.cpp:
```diff
@@ -36,9 +36,24 @@
 namespace WebCore {
 
-static CheckedUint32 computeDataSize(const IntSize& size)
+static PixelFormat toPixelFormat(ImageDataStorageFormat storageFormat)
```
Why is this function here? Should not be implemented as inline function in ImageDataStorageFormat.h?

shallawa commented on Source/WebCore/html/ImageData.cpp:
```diff
@@ -48,6 +63,11 @@ PredefinedColorSpace ImageData::computeColorSpace(std::optional<ImageDataSetting
     return defaultColorSpace;
 }
 
+static ImageDataStorageFormat computeStorageFormat(std::optional<ImageDataSettings> settings, ImageDataStorageFormat defaultStorageFormat = ImageDataStorageFormat::Uint8)
+{
+    return settings ? settings->storageFormat : defaultStorageFormat;
+}
```
The optional `ImageDataSettings` and the default values are a little bit confusing in `computeStorageFormat()` and `computeColorSpace()`. For example this call does not make sense to me.
```
auto computedColorSpace = ImageData::computeColorSpace(settings, m_settings.colorSpace);
```
Why is this a method of `ImageData`? Should not it be a constructor or a function in `ImageDataSettings`?

squelart replied:
`computeColorSpace` predates this patch, and I followed in its path.
It makes sense to me: The optional settings comes from the JS API (like `getImageData()`), where the caller may or may not provide an ImageDataSettings dictionary that may or may not contain properties like colorSpace and StorageFormat.

In that call you're mentioning, `m_settings.colorSpace` is the *canvas* pixel format, and it's used as the default if the caller of `getImageData` doesn't provide an overriding value.
`ImageDataSettings` itself doesn't know about canvases, it's just a dumb struct corresponding to the JS `dictionary ImageDataSettings`.

shallawa commented on Source/WebCore/html/ImageData.cpp:
```diff
@@ -61,41 +81,52 @@ RefPtr<ImageData> ImageData::create(RefPtr<ByteArrayPixelBuffer>&& pixelBuffer)
 }
 
-RefPtr<ImageData> ImageData::create(const IntSize& size, PredefinedColorSpace colorSpace)
+RefPtr<ImageData> ImageData::createZeroFilled(const IntSize& size, PredefinedColorSpace colorSpace, ImageDataStorageFormat storageFormat)
```
Why did you change the name? Does `ImageData` have a create method which does not fill array with zeros?

shallawa replied:
I think this function should take a last parameter `settings` which should be defaulted to `std::nullopt`.

squelart replied:
- Name change: It was a remnant of previously having createUninitialized, now removed, so I'll revert this name to `create`.
- Optional settings: Actually not used anywhere, so no need to add it now, we can add it if/when needed.
Or do you see value in having the optional settings before we actually need it?

shallawa commented on Source/WebCore/html/ImageData.cpp:
```diff
@@ -61,41 +81,52 @@ RefPtr<ImageData> ImageData::create(RefPtr<ByteArrayPixelBuffer>&& pixelBuffer)
 }
 
-ExceptionOr<Ref<ImageData>> ImageData::createUninitialized(unsigned rows, unsigned pixelsPerRow, PredefinedColorSpace defaultColorSpace, std::optional<ImageDataSettings> settings)
+ExceptionOr<Ref<ImageData>> ImageData::createFromBufferOrZero(unsigned sw, unsigned sh, PredefinedColorSpace defaultColorSpace, std::optional<ImageDataSettings> settings, const uint8_t* buffer, size_t length)
```
I think `settings` should be the last parameter and should be defaulted to `std::nullopt`.

shallawa commented on Source/WebCore/html/ImageData.cpp:
```diff
@@ -61,41 +81,52 @@ RefPtr<ImageData> ImageData::create(RefPtr<ByteArrayPixelBuffer>&& pixelBuffer)
+ExceptionOr<Ref<ImageData>> ImageData::createZeroFilled(unsigned sw, unsigned sh, PredefinedColorSpace defaultColorSpace, std::optional<ImageDataSettings> settings)
+{
+    return createFromBufferOrZero(sw, sh, defaultColorSpace, settings, nullptr, 0);
 }
```
I think these functions will not be needed if the other `createFromBufferOrZero()` makes its settings parameter be the last and the other `createZeroFilled()` adds a new settings parameter.

shallawa commented on Source/WebCore/html/ImageData.h:
```diff
@@ -43,9 +43,11 @@ class ImageData : public RefCounted<ImageData> {
     WEBCORE_EXPORT static Ref<ImageData> create(Ref<ByteArrayPixelBuffer>&&);
     WEBCORE_EXPORT static RefPtr<ImageData> create(RefPtr<ByteArrayPixelBuffer>&&);
-    WEBCORE_EXPORT static RefPtr<ImageData> create(const IntSize&, PredefinedColorSpace);
+    WEBCORE_EXPORT static RefPtr<ImageData> createZeroFilled(const IntSize&, PredefinedColorSpace, ImageDataStorageFormat = ImageDataStorageFormat::Uint8);
```
I think this should take `ImageDataSettings` instead of `PredefinedColorSpace, ImageDataStorageFormat = ImageDataStorageFormat::Uint8`

squelart replied:
I'm removing ImageDataStorageFormat, never used.
If it's needed in a later patch, I'll consider changing to ImageDataSettings, thanks for the suggestion...

shallawa commented on Source/WebCore/html/ImageData.h:
```diff
@@ -43,9 +43,11 @@ class ImageData : public RefCounted<ImageData> {
     WEBCORE_EXPORT static RefPtr<ImageData> create(const IntSize&, ImageDataArray&&, PredefinedColorSpace);
-    WEBCORE_EXPORT static ExceptionOr<Ref<ImageData>> createUninitialized(unsigned rows, unsigned pixelsPerRow, PredefinedColorSpace defaultColorSpace, std::optional<ImageDataSettings> = std::nullopt);
+
+    WEBCORE_EXPORT static ExceptionOr<Ref<ImageData>> createZeroFilled(unsigned sw, unsigned sh, PredefinedColorSpace defaultColorSpace, std::optional<ImageDataSettings> = std::nullopt);
```
I think this should take `ImageDataSettings` instead of `PredefinedColorSpace defaultColorSpace, std::optional<ImageDataSettings> = std::nullopt`

squelart replied:
I've tried that, but:
- Some generated JS binding code provides an `optional< ImageDataSettings>`, so we must still accept that.
- `ImageDataSettings::colorSpace` is itself `optional`, so every caller would have to ensure it's provided if needed.
- The behavior of defaultColorSpace+optional settings is given in the specs, so I think it's good to have it centralized in one place, rather than trying to replicate it at multiple call sites.
- Because of the above, there were a few places that needed some fiddly code, something like `{ { colorSpace } }`, or worse:
```
        if (!settings)
            settings = { { m_settings.colorSpace } };
        else if (!settings->colorSpace)
            settings->colorSpace = m_settings.colorSpace;
```

So I've gone with:
- One `Create` with both,
- and one `Create` with just `optional<ImageDataSettings>`, I think it's clearer at all call sites (and bonus: the implementation is trivial).

shallawa commented on Source/WebCore/html/ImageDataArray.cpp:
```diff
@@ -29,6 +29,185 @@
+
+std::optional<ImageDataArray> ImageDataArray::tryCreateFromBufferOrZero(size_t length, ImageDataStorageFormat storageFormat, const uint8_t* buffer, size_t bytes)
+{
+    std::optional<ImageDataArray> array = tryCreate(length, storageFormat, false);
```
Should we check `length >= bytes`?

shallawa commented on Source/WebCore/html/ImageDataArray.cpp:
```diff
@@ -29,6 +29,185 @@
+    case JSC::TypeFloat16: return ImageDataStorageFormat::Float16;
+    default: RELEASE_ASSERT_NOT_REACHED("Unexpected ArrayBufferView type");
+    }
+}
```
Should the conversion from `TypedArrayType` to `ImageDataStorageFormat` be in ImageDataStorageFormat.h?

shallawa commented on Source/WebCore/html/ImageDataArray.cpp:
```diff
@@ -29,6 +29,185 @@
+    });
+}
+
+Ref<JSC::Uint8ClampedArray> ImageDataArray::asUint8ClampedArray() const
+{
+    return convertToTypedArray<JSC::Uint8ClampedArray>(*this);
+}
+
+Ref<JSC::Float16Array> ImageDataArray::asFloat16Array() const
+{
+    return convertToTypedArray<JSC::Float16Array>(*this);
+}
```
This interface is now ambiguous. The return value might be just a reference to the `ArrayBufferView` of the `ImageData` if no conversion happened. But it might be a new array converted from the existing `ArrayBufferView` if the requested format is not the same as the format of the `ImageData`. 

Is this covered by the specs? What should happen if JS modifies the converted array? Should this affect the stored array?

For example you create a Float16 ImageData and you requested a Uint8 array. Then you modified this Uint8 array. Should this modification be reflected in the stored ImageData Float16 array?

squelart replied:
From the JS API perspective, the user would use the `data` property or `ImageData`, which directly provides the `(Uint8ClampedArray or Float16Array)` object, no issue there.

`asUint8ClampedArray` effectively replaces the old **internal** `ImageData::data()` that used to always return a `Ref<Uint8ClampedArray>`, so at least the uint8-everywhere version (the default with the feature OFF) should work just like before.
And I believe these should go away once float16 is fully supported.

But for now, as far as I can see, all its uses are as a source of read-only pixels (for blobs, dataURL, put's...).
And in some of these cases, we'd want to avoid copies when not needed, and it would probably have a noticeable perf impact.

Happy to discuss further... Better naming maybe?

shallawa replied:
You are right. I think I was confused. I thought both of the formats will be exposed to JS. But I think the created format will be exposed only to JS. The asUint8ClampedArray() will be used by `putImageData()` for now. right?

shallawa commented on Source/WebCore/html/ImageDataArray.h:
```diff
@@ -28,34 +28,48 @@
+    size_t length() const;
+
+    JSC::ArrayBufferView& arrayBufferView() const { return m_arrayBufferView.get(); }
+    JSC::ArrayBufferView* operator->() const { return &arrayBufferView(); }
```
This operator is weird. It makes ImageDataArray act like a pointer. Usually when this operator is implemented, the `operator*` and the `operator bool` are also implemented. So if your class allows you to say `foo->bar`, it should allow you to say `*foo` and `if (foo)`. But this does not happen here. `std::optional` is an example.

squelart replied:
I'll remove it, it shouldn't be needed anymore after I've added pass-through functions like `span()` as you suggested/requested.
And if there's still a need to access the `ArrayBufferView`, `arrayBufferView()` is there for that.

shallawa commented on Source/WebCore/html/ImageDataArray.cpp:
```diff
@@ -29,6 +29,165 @@
+    ASSERT(isSupported(m_arrayBufferView.get()));
+}
+
+ImageDataArray::ImageDataArray(Ref<JSC::Uint8ClampedArray>&& data)
+    : ImageDataArray(Ref<JSC::ArrayBufferView>(WTFMove(data)))
+{ }
+
+ImageDataArray::ImageDataArray(Ref<JSC::Float16Array>&& data)
+    : ImageDataArray(Ref<JSC::ArrayBufferView>(WTFMove(data)))
+{ }
```
Do we really need these two constructors? Since `ArrayBufferView` is the base class of `Uint8ClampedArray` and `Float16Array`, I think the constructor that takes `Ref<JSC::ArrayBufferView>&&` should be sufficient.

squelart replied:
The issue is that a random `ArrayBufferView` could be other than uint8 or float16, which we don't support.
These two constructors cannot fail, and they ensure that `ImageDataArray` is always valid by construction.

An alternative would be to have something like `static std::optional<ImageDataArray> tryCreate(Ref<JSC::ArrayBufferView>&&)`, but I think it would make life unnecessarily more difficult for callers that already have a `Uint8ClampedArray` or `Float16Array` at hand.

So I'll keep these for now, but we can always revisit if something else becomes more appropriate as development continues...

shallawa commented on Source/WebCore/html/ImageDataArray.cpp:
```diff
@@ -29,6 +29,165 @@
+static void fillTypedArray(TypedArray& typedArray, std::span<const uint8_t> optionalBytes)
+{
+    if (optionalBytes.empty())
+        typedArray.zeroFill();
```
Early return?

shallawa commented on Source/WebCore/html/ImageData.cpp:
```diff
@@ -63,83 +68,75 @@ RefPtr<ImageData> ImageData::create(RefPtr<ByteArrayPixelBuffer>&& pixelBuffer)
+    if (!array)
         return nullptr;
-    return adoptRef(*new ImageData(size, byteArray.releaseNonNull(), colorSpace));
+    return adoptRef(*new ImageData(size, WTFMove(*array), colorSpace));
```
Is this faster than `releaseNonNull()`? I see you are using `typedArray.releaseNonNull()` in another place.

squelart replied:
In the old code, `byteArray` was a `Ref<Uint8ClampedArray>`, so `releaseNonNull()` was the best choice then.
But now, `array` here is an `std::optional<ImageDataArray>`, so we do need to use `*` to access the optional's value, and `WTFMove` to make it an rvalue-reference as expected by the `ImageData` constructor.

squelart committed:
ImageData using storageFormat
https://bugs.webkit.org/show_bug.cgi?id=284004
rdar://problem/140877760

Reviewed by Said Abou-Hallawa.

ImageData now supports:
- The IDL property `storageFormat` of type `ImageDataStorageFormat`, which
  reflects the ImageDataSettings.storageFormat value given during construction.
- The IDL property `data` can now be `Uint8ClampedArray` or `Float16Array`,
  depending on the storageFormat.

The main difference for now, is that
`context.createImageData(w, h, { storageFormat: "float16" })`
should correctly output an ImageData with data of type Float16Array.

Otherwise most uses of ImageData, like in CanvasRenderingContext2DBase, enforce
the use of Uint8ClampedArray, so there shouldn't be any visible changes in most
cases, including when requesting
`context.getImageData(w, h, { storageFormat: "float16" })`
(to be implemented in future PRs).

* LayoutTests/fast/canvas/imagedata-storageformat-enabled-expected.txt:
* LayoutTests/fast/canvas/imagedata-storageformat-enabled.html:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/bindings/js/JSImageDataCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::CloneDeserializer::readTerminal):
* Source/WebCore/html/ImageData.cpp:
(WebCore::computeDataSize):
(WebCore::computeStorageFormat):
(WebCore::ImageData::create):
(WebCore::ImageData::pixelBuffer const):
(WebCore::ImageData::createUninitialized): Deleted.
(WebCore::ImageData::clone const): Deleted.
* Source/WebCore/html/ImageData.h:
(WebCore::ImageData::create):
(WebCore::ImageData::data const):
(WebCore::ImageData::storageFormat const):
* Source/WebCore/html/ImageData.idl:
* Source/WebCore/html/ImageDataArray.cpp:
(isType):
(WebCore::visitArrayBufferView):
(WebCore::ImageDataArray::ImageDataArray):
(WebCore::fillTypedArray):
(WebCore::ImageDataArray::tryCreate):
(WebCore::ImageDataArray::isSupported):
(WebCore::ImageDataArray::storageFormat const):
(WebCore::convertToTypedArray):
(WebCore::ImageDataArray::asUint8ClampedArray const):
(WebCore::ImageDataArray::asFloat16Array const):
(WebCore::ImageDataArray::length const):
(WebCore::ImageDataArray::copyToJSONArray const):
(WebCore::ImageDataArray::operator Variant const):
* Source/WebCore/html/ImageDataArray.h:
(WebCore::ImageDataArray::isSupported):
(WebCore::ImageDataArray::tryCreate):
(WebCore::ImageDataArray::arrayBufferView const):
(WebCore::ImageDataArray::protectedArrayBufferView const):
(WebCore::ImageDataArray::byteLength const):
(WebCore::ImageDataArray::isDetached const):
(WebCore::ImageDataArray::span const):
(WebCore::ImageDataArray::ImageDataArray): Deleted.
(WebCore::ImageDataArray::length const): Deleted.
(WebCore::ImageDataArray::get const): Deleted.
(WebCore::ImageDataArray::data const): Deleted.
* Source/WebCore/html/ImageDataStorageFormat.h:
(WebCore::toPixelFormat):
(WebCore::toImageDataStorageFormat):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::createImageData const):
(WebCore::CanvasRenderingContext2DBase::cacheImageDataIfPossible):
(WebCore::CanvasRenderingContext2DBase::makeImageDataIfContentsCached const):
(WebCore::CanvasRenderingContext2DBase::getImageData const):
(WebCore::initializeEmptyImageData): Deleted.
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::texImageSource):
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::buildArrayForImageData):

Canonical link: https://commits.webkit.org/287598@main
